node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -205,16 +227,102 @@ impl OrderingProvider {
 			self.included_candidates.remove(&finalized_candidate);
 		}
 	}
+
+	/// Returns ancestors of `head` in the descending order, stopping
+	/// either at the block present in cache or the latest finalized block.
+	///
+	/// Suited specifically for querying non-finalized chains, thus
+	/// doesn't rely on block numbers.
+	///
+	/// Both `head` and last are **not** included in the result.
+	async fn get_block_ancestors<Sender: SubsystemSender>(
+		&mut self,
+		sender: &mut Sender,
+		mut head: Hash,
+		mut head_number: BlockNumber,
+	) -> NonFatalResult<Vec<Hash>> {
+		const ANCESTRY_STEP: usize = 10;
+
+		let mut ancestors = Vec::new();
+
+		if self.last_observed_blocks.get(&head).is_some() {
+			return Ok(ancestors)
+		}
+
+		let finalized_block_number = if let Ok(number) = get_finalized_block_number(sender).await {
+			number
+		} else {
+			return Ok(ancestors)
+		};
+
+		loop {
+			let (tx, rx) = oneshot::channel();
+			let hashes = {
+				sender
+					.send_message(
+						ChainApiMessage::Ancestors {
+							hash: head,
+							k: ANCESTRY_STEP,
+							response_channel: tx,
+						}
+						.into(),
+					)
+					.await;
+
+				rx.await??
+			};
+
+			let earliest_block_number = head_number - hashes.len() as u32;
+			// The reversed order is parent, grandparent, etc. excluding the head.
+			let block_numbers = (earliest_block_number..head_number).rev();
+
+			for (block_number, hash) in block_numbers.zip(&hashes) {
+				if self.last_observed_blocks.get(&head).is_some() ||
+					block_number == finalized_block_number
```

Comment:
<= would make this function a bit more robust in case a `head_number` < `finalized_block_number` would get passed in.